[WIP]feat(component): Added Date Time Picker component

MariaAga opened the pull request:
**What**: Add the DateTimePicker component from PF to PF-react.

The PatternFly design is at:https://www.patternfly.org/pattern-library/forms-and-controls/date-and-time/#design

Currently there is only a Date Picker

TheRealJon commented on packages/patternfly-3/patternfly-react/src/components/DateTimePicker/Components/PickTimeClock.js:
```diff
@@ -0,0 +1,100 @@
+    const minutes = time.getMinutes();
+    let hours = this.props.time.getHours();
+    const ampm = hours > 12 ? 'PM' : 'AM';
+    hours %= 12;
```
Just make this calculation where `hours` is evaluated below.
```suggestion
```

TheRealJon commented on packages/patternfly-3/patternfly-react/src/components/DateTimePicker/Components/PickTimeClock.js:
```diff
@@ -0,0 +1,100 @@
+              <tr>
+                <td onClick={() => toggleTimeTable('H')}>
+                  <span className="timepicker-hours" title="Pick Hour">
+                    {hours}
```
```suggestion
                    {hours%12}
```

TheRealJon commented on packages/patternfly-3/patternfly-react/src/components/DateTimePicker/Components/PickTimeTable.js:
```diff
@@ -0,0 +1,53 @@
+
+class PickTimeTable extends React.Component {
+  setTime = (newTime, type) => {
+    const { time } = this.props;
```
```suggestion
    const { time, setSelected } = this.props;
```

TheRealJon commented on packages/patternfly-3/patternfly-react/src/components/DateTimePicker/Components/PickTimeTable.js:
```diff
@@ -0,0 +1,53 @@
+    const { time } = this.props;
+    if (type === 'M') time.setMinutes(newTime);
+    else if (type === 'H') time.setHours(newTime);
+    this.props.setSelected(time);
```
```suggestion
    setSelected(time);
```

TheRealJon commented on packages/patternfly-3/patternfly-react/src/components/DateTimePicker/DatePicker.js:
```diff
@@ -0,0 +1,24 @@
+import PropTypes from 'prop-types';
+import Month from './Calendar/Month';
+
+// eslint-disable-next-line react/prefer-stateless-function
```
This component could be a React SFC
```jsx
const DatePicker => ({date, setSelected}) => <div className="datepicker">
  <Month date={date} setSelected={setSelected} />
</div>;
```

TheRealJon commented on packages/patternfly-3/patternfly-react/src/components/DateTimePicker/Components/TodayButton.js:
```diff
@@ -0,0 +1,32 @@
+
+class TodayButton extends React.Component {
+  setToday = () => {
+    this.props.setSelected(new Date());
```
Should do a null check for `setSelected`prop.

TheRealJon commented on packages/patternfly-3/patternfly-react/src/components/DateTimePicker/TimePicker.js:
```diff
@@ -0,0 +1,47 @@
+    isTimeTableOpen: false
+  };
+
+  toggleTimeTable = (type = null) => {
```
Since we default the type to `null`, this could probably be simplified to:

```js
toggleTimeTable = (type = null) => {
  this.setState({
    type,
    isTimeTableOpen: !this.state.isTimeTableOpen
  });
}
```

TheRealJon commented on packages/patternfly-3/patternfly-react/src/components/DateTimePicker/TimePicker.js:
```diff
@@ -0,0 +1,47 @@
+    else this.setState({ isTimeTableOpen: false });
+  };
+  render() {
+    const { time } = this.state;
```
```suggestion
    const { isTimeTableOpen, time, typeOfTimeTable } = this.state;
    const { setSelected } = this.props;
```

TheRealJon commented on packages/patternfly-3/patternfly-react/src/components/DateTimePicker/TimePicker.js:
```diff
@@ -0,0 +1,47 @@
+          <PickTimeTable
+            time={time}
+            setSelected={this.props.setSelected}
+            type={this.state.typeOfTimeTable}
```
```suggestion
            type={typeOfTimeTable}
```

TheRealJon commented on packages/patternfly-3/patternfly-react/src/components/DateTimePicker/TimePicker.js:
```diff
@@ -0,0 +1,47 @@
+            time={time}
+            setSelected={this.props.setSelected}
+            type={this.state.typeOfTimeTable}
+            show={this.state.isTimeTableOpen}
```
```suggestion
            show={isTimeTableOpen}
```

TheRealJon commented on packages/patternfly-3/patternfly-react/src/components/DateTimePicker/TimePicker.js:
```diff
@@ -0,0 +1,47 @@
+        {this.state.isTimeTableOpen ? (
+          <PickTimeTable
+            time={time}
+            setSelected={this.props.setSelected}
```
```suggestion
            setSelected={setSelected}
```

TheRealJon commented on packages/patternfly-3/patternfly-react/src/components/DateTimePicker/TimePicker.js:
```diff
@@ -0,0 +1,47 @@
+            show={this.state.isTimeTableOpen}
+          />
+        ) : (
+          <PickTimeClock time={time} setSelected={this.props.setSelected} toggleTimeTable={this.toggleTimeTable} />
```
```suggestion
          <PickTimeClock time={time} setSelected={setSelected} toggleTimeTable={this.toggleTimeTable} />
```

MariaAga commented:
@TheRealJon Thanks for all the comment! The bug you described is still happening and will be fixed.
I've added a Time Picker, but this is still a WIP so the code is not completely 'clean'

Laviro commented:
@patternfly/patternfly-react-ux could you have a look? thanks

Laviro commented on packages/patternfly-3/patternfly-react/src/components/DateTimePicker/TimeComponents/PickTimeClock.js:
```diff
@@ -0,0 +1,103 @@
+    return (
+      <div className="timepicker-picker">
+        <table>
+          <tr>
```
Please add a `tbody` here as the browser throws an error about it:
![selection_024](https://user-images.githubusercontent.com/26363699/53696402-4106cc80-3dcf-11e9-910f-b9502cdde4aa.png)

amirfefer replied:
Insert `tbody` will break patternfly css 
But according to [react](https://github.com/facebook/react/issues/5652)  it is  important to have this `tbody`

how about moving the fixed css file to patternfly-react until `patternfly` will fix it in patternfly-core?

MariaAga replied:
added tbody

Laviro commented on packages/patternfly-3/patternfly-react/src/components/DateTimePicker/TimeComponents/PickTimeClock.js:
```diff
@@ -0,0 +1,103 @@
+        <table>
+          <tr>
+            <td onClick={() => this.setTime('H', 1)}>
+              <a href="#" title="Increment Hour" className="btn clock-btn">
```
because you used `href="#"`
when clicking on the time-picker buttons, it reloads the page or take you to another page:

![storybook 2](https://user-images.githubusercontent.com/26363699/53696488-03567380-3dd0-11e9-9ff8-8507416c398a.gif)

Laviro commented on packages/patternfly-3/patternfly-react/src/components/DateTimePicker/DatePicker.js:
```diff
@@ -0,0 +1,72 @@
+  state = {
+    value: new Date(this.props.value)
+  };
+  formatDate = () => {
```
this function has been used in several places over the code, 
could you move it into an `helper` file and extract it from there ?

MariaAga replied:
Its different in every place (one for date, one for time, one for date-time) , and exporting it and adding its different variations will make it too complicated in my opinion

Laviro replied:
got it.. thanks

Laviro commented on packages/patternfly-3/patternfly-react/src/components/DateTimePicker/DateComponents/DateInput.js:
```diff
@@ -0,0 +1,93 @@
+    const { date, locale, weekStartsOn, className, setSelected } = this.props;
+    const { typeOfDateInput } = this.state;
+    return (
+      <div className={`datepicker ${className}`}>
```
could you use the `classname` library ?
```
import cx from 'classnames';
...
<div className={cx(datepicker, className)}>
```

Laviro commented on packages/patternfly-3/patternfly-react/src/components/DateTimePicker/DateComponents/DateInput.js:
```diff
@@ -0,0 +1,93 @@
+    return (
+      <div className={`datepicker ${className}`}>
+        {(() => {
+          switch (typeOfDateInput) {
```
could be moved to an helper file into a function like: `getDateViewByType` ?

MariaAga replied:
I moved it into a function. I think it should stay in the DateInput file and shouldn't be in a helper file because only the DateInput component will use it.

Laviro commented on packages/patternfly-3/patternfly-react/src/components/DateTimePicker/DateComponents/Day.js:
```diff
@@ -0,0 +1,40 @@
+class Day extends React.Component {
+  onClick = e => {
+    e.preventDefault();
+    this.props.setSelected(this.props.day);
```
Could you use destruction here?
```
const { setSelected, day } = this.props
setSelected(day)
```

Laviro commented on packages/patternfly-3/patternfly-react/src/components/DateTimePicker/DateComponents/Day.js:
```diff
@@ -0,0 +1,40 @@
+import classNames from 'classnames';
+import { isEquelDate, isWeekend } from './helpers';
+
+class Day extends React.Component {
```
I think we can use a simple component here instead of a class

Laviro commented on packages/patternfly-3/patternfly-react/src/components/DateTimePicker/DateComponents/Day.js:
```diff
@@ -0,0 +1,40 @@
+  };
+  render() {
+    const day = this.props.day.getDate();
+    const { day: date, currDate, selectedDate } = this.props;
```
maybe:
```
const { day, currDate, selectedDate } = this.props;
const date = day.getDate();
```

Laviro commented on packages/patternfly-3/patternfly-react/src/components/DateTimePicker/DateComponents/Day.js:
```diff
@@ -0,0 +1,40 @@
+      old: date.getMonth() !== currDate.getMonth(),
+      active: isEquelDate(date, selectedDate),
+      today: isEquelDate(date, now)
+    });
```
optional: could this component receive all of the props ready from its parent,
instead of calculation everything here?

Laviro commented on packages/patternfly-3/patternfly-react/src/components/DateTimePicker/DateComponents/DecadeView.js:
```diff
@@ -0,0 +1,85 @@
+  setSelectedYear = year => {
+    const { date } = this.state;
+    date.setFullYear(year);
+    this.props.setSelected(date);
```
could you use destruction here?

Laviro commented on packages/patternfly-3/patternfly-react/src/components/DateTimePicker/DateComponents/DecadeView.js:
```diff
@@ -0,0 +1,85 @@
+        <table className="table-condensed">
+          <thead>
+            <tr>
+              <th className="prev" onClick={this.getPrevDecade}>
```
optional, maybe divide it into smaller components?

MariaAga replied:
Divided into DecadeViewHeader and DecadeViewTable

Laviro commented on packages/patternfly-3/patternfly-react/src/components/DateTimePicker/DateComponents/Header.js:
```diff
@@ -0,0 +1,68 @@
+
+class Header extends React.Component {
+  getWeekStart = date => addDays(date, (7 - date.getDay()) % 7);
+  weekArray = () => {
```
as suggested on other places, maybe move this logic into helpers and keep this component simple.

Laviro commented on packages/patternfly-3/patternfly-react/src/components/DateTimePicker/DateComponents/TodayButton.js:
```diff
@@ -0,0 +1,33 @@
+        <tbody>
+          <tr>
+            <td>
+              <a onClick={this.setToday} title="Go to today">
```
instead of an anchor `<a />`  you should use a button as this isn't a link, [read more](https://bitsofco.de/anchors-vs-buttons/)

Laviro commented on packages/patternfly-3/patternfly-react/sass/patternfly-react/_date-time-picker.scss:
```diff
@@ -0,0 +1,22 @@
+    padding-bottom: 4px;
+    padding-top: 4px;
+  }
+}
```
I saw that there are small places where the elements are a bit off position, but I guess that the designers will have more to say about it than myself:

**Current implementation:**

![date-picker](https://user-images.githubusercontent.com/26363699/53699722-124f1d00-3df4-11e9-99fc-74d3adee5d27.png)

**Original Design:**

![date-picker-design](https://user-images.githubusercontent.com/26363699/53699724-1b3fee80-3df4-11e9-95c4-0322627758ef.png)

amirfefer commented:
one more thing, can you hide the popover when an outside click occurs ?

MariaAga commented:
@amirfefer added inputProps to the datetimepicker input, and now the popover gets hidden when an outside click occurs

amirfefer commented on packages/patternfly-3/patternfly-react/sass/patternfly-react/_date-time-picker.scss:
```diff
@@ -0,0 +1,33 @@
+.input-group.date-time-picker-pf.input-group {
+  width: 41.66666667%;
```
why is it needed? 
it cause incompatibility with other inputs in a form

mcarrano commented:
@MariaAga I am not seeing a preview link. Can you generate one so we can perform a UX review? Thanks.

Laviro commented:
@MariaAga This is how to deploy manually:

1. yarn build:storybook && cp -r .out .public/patternfly-3

2. yarn build:prdocs && cp -r packages/patternfly-4/react-docs/public .public/patternfly-4 && cp -r packages/patternfly-4/react-docs/public/assets .public/assets

3. yarn run surge --project .public --domain $DEPLOY_DOMAIN

you may need to register into surge before.

MariaAga commented:
Thanks Ron, added a [preview ](https://1420-pr-patternfly-react-patternfly.surge.sh/patternfly-3/index.html?selectedKind=patternfly-react%2FForms%20and%20Controls%2FDate%20Time%20Picker&selectedStory=DateTimePicker&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybooks%2Fstorybook-addon-knobs)
@mcarrano

mcarrano commented:
This looks great @MariaAga . The only design thing I see is the positioning of the Today link. Can this be centered below the calendar? I did notice that you used the European style calendar with the week starting on Monday rather than Sunday like the core implementation. Is that something we can set as a Prop so the user can decide what calendar style to use?

I also felt it would be useful to include variants that either only have the date picker or the time picker by themselves.

patternfly-build commented:
PatternFly-React preview: https://1420-pr-patternfly-react-patternfly.surge.sh

amirfefer commented on packages/patternfly-3/patternfly-react/src/components/DateTimePicker/DateTimePicker.js:
```diff
@@ -0,0 +1,84 @@
+    );
+    return (
+      <div>
+        <OverlayTrigger trigger="click" placement="top" overlay={popover} rootClose>
```
You can use `target` prop in order to refer popover's position  to other element

amirfefer commented on packages/patternfly-3/patternfly-react/src/components/DateTimePicker/DateTimePicker.js:
```diff
@@ -0,0 +1,84 @@
+    const { locale, weekStartsOn, inputProps } = this.props;
+    const { value, typeOfDateInput, isTimeTableOpen } = this.state;
+    const popover = (
+      <Popover id="popover-date-picker" className="bootstrap-datetimepicker-widget dropdown-menu timepicker-sbs">
```
hard coded id doesn't allow multiple usage, can be prop

amirfefer commented on packages/patternfly-3/patternfly-react/src/components/DateTimePicker/DateComponents/TodayButton.js:
```diff
@@ -0,0 +1,33 @@
+        <tbody>
+          <tr>
+            <td>
+              <button className="today-button" onClick={this.setToday} title="Go to today">
```
please add ` type="button"`, without a type in forms, clicking on this  would try to submit

MariaAga replied:
Added, thanks

rvsia commented on packages/patternfly-3/patternfly-react/src/components/DateTimePicker/DateComponents/DecadeViewHeader.js:
```diff
@@ -0,0 +1,31 @@
+  <thead>
+    <tr>
+      <th className="prev" onClick={getPrevDecade}>
+        <span className="glyphicon glyphicon-chevron-left" title="Previous Decade" />
```
I think all titles should be accessible (because of localization) to a consumer.

MariaAga replied:
titles were used as tool tips and are now removed

amirfefer commented on packages/patternfly-3/patternfly-react/src/components/DateTimePicker/DateTimePicker.js:
```diff
@@ -0,0 +1,84 @@
+              {...inputProps}
+              aria-label="date-picker-input"
+              type="text"
+              value={this.formatDate()}
```
Date Time input value should be [controlled](https://reactjs.org/docs/forms.html) and support nul (doesn't have to be required in forms)

amirfefer commented on packages/patternfly-3/patternfly-react/src/components/DateTimePicker/TimePicker.js:
```diff
@@ -0,0 +1,84 @@
+  render() {
+    const { locale, weekStartsOn } = this.props;
+    const popover = (
+      <Popover id="popover-date-picker" className="bootstrap-datetimepicker-widget dropdown-menu">
```
what would happen if there are multiple date time components in a page?

amirfefer commented on packages/patternfly-3/patternfly-react/src/components/DateTimePicker/TimePicker.js:
```diff
@@ -0,0 +1,84 @@
+  state = {
+    value: this.getDateFromTime(this.props.value)
+  };
+  formatDate = () => {
```
can date format support an optional prop ? (i.e 'yyyy-mm-dd HH:MM:SS`), in some cases the user needs to bypass localization.

amirfefer commented on packages/patternfly-3/patternfly-react/src/components/DateTimePicker/TimePicker.js:
```diff
@@ -0,0 +1,84 @@
+            <FormControl
+              aria-label="date-time-picker-input"
+              type="text"
+              value={this.formatDate()}
```
this value is presented to the user and also sent to the server, wouldn't it be better to separate between the two?

amirfefer commented on packages/patternfly-3/patternfly-react/src/components/DateTimePicker/TimePicker.js:
```diff
@@ -0,0 +1,84 @@
+    );
+    return (
+      <div>
+        <OverlayTrigger trigger="click" placement="top" overlay={popover} rootClose>
```
Popover position needs to be calculated or be a prop

patternfly-build commented:
PatternFly-React preview: https://1420-pr-patternfly-react-patternfly.surge.sh

patternfly-build commented:
PatternFly-React preview: https://1420-pr-patternfly-react-patternfly.surge.sh

MariaAga commented:
@jeff-phillips-18 I'm having some trouble with the css, and can't get  the Time Picker to look like the design, can you help me with it? 
The popover takes the entire screen width, the picker itself only takes half the space of the popover, and in the design the blue line in the clock is smaller then it is in the demo.
[Design]( https://www.patternfly.org/pattern-library/forms-and-controls/date-and-time/#design)
[Demo](https://1420-pr-patternfly-react-patternfly.surge.sh/?knob-Week%20Start=1&selectedKind=patternfly-react%2FForms%20and%20Controls%2FDate%20Time%20Picker&selectedStory=DateTimePicker&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybooks%2Fstorybook-addon-knobs)

jeff-phillips-18 commented:
To fix those issues. try:

https://github.com/patternfly/patternfly-react/pull/1420/files#diff-0059075e9ea58c01ca3a492af0525f08R46
Remove `className="col-md-6"`

is the _date-time-picker.scss :
https://github.com/patternfly/patternfly-react/pull/1420/files#diff-933694b69d83cffdc64e5ed9930ee16aR19
add:
```
  table td {
    height: initial;
    line-height: initial;
    width: initial;
```

within the .bootstrap-datetimepicker-widget block, add:
```
  &.bootstrap-timepicker-widget {
    width: 16em;
    .popover-content {
      padding-left: 0;
    }
  }
```

Hope this helps, these show other issues with placement as well but that gets more complicated. Basically you can't set `left: 0 !important;` in `.datepicker-widget` since that will only work if the time picker component is set to the far left of the window as it is in the demo.

MariaAga commented:
Thanks @jeff-phillips-18, I will try it.
The `left: 0 !important;` was pre-set by Patternfly main project (not Patternfly-react) and not by me

jeff-phillips-18 commented:
Not sure how Patternfly was using it but in the popover, the position is `absolute` so left: 0 will always be far left.

MariaAga commented:
@jeff-phillips-18 Thanks for your help, the timepicker looks much better!

patternfly-build commented:
PatternFly-React preview: https://1420-pr-patternfly-react-patternfly.surge.sh

jeff-phillips-18 commented on packages/patternfly-3/patternfly-react/sass/patternfly-react/_date-time-picker.scss:
```diff
@@ -0,0 +1,56 @@
+.bootstrap-datetimepicker-widget.top {
```
this should have a -pf suffix somewhere. Probably add another class like `date-time-picker-pf` and have the `.bootstrap-datetimepicker-widget` inside of that, i.e.:
```
.date-time-picker-pf {
  &.bootstrap-datetimepicker-widget {
...
```

This will scope it to only those bootstrap date time picker components that are indeed the pf-react version. Some apps could use a mix 🤷‍♂️

MariaAga replied:
Added, thanks!

jeff-phillips-18 commented on packages/patternfly-3/patternfly-react/sass/patternfly-react/_date-time-picker.scss:
```diff
@@ -0,0 +1,56 @@
+  }
+}
+
+.datepicker-widget {
```
Should have a `-pf` suffix

jeff-phillips-18 commented on packages/patternfly-3/patternfly-react/sass/patternfly-react/_date-time-picker.scss:
```diff
@@ -0,0 +1,56 @@
+  .arrow {
+    left: 10% !important;
+  }
+}
```
Please add the .less version as well as we still support both.

jeff-phillips-18 commented on packages/patternfly-3/patternfly-react/src/components/DateTimePicker/DateComponents/helpers.js:
```diff
@@ -0,0 +1,40 @@
+  return result;
+};
+
+export const isEquelDate = (date1, date2) =>
```
typo -> `isEqualDate`

MariaAga replied:
fixed, thanks

jeff-phillips-18 commented on packages/patternfly-3/patternfly-react/src/components/DateTimePicker/DateComponents/DateInput.test.js:
```diff
@@ -0,0 +1,40 @@
+    .first()
+    .simulate('click');
+  expect(component.render()).toMatchSnapshot();
+});
```
Please add a test for an invalid date format.

MariaAga replied:
Added a test in DateInput and TimeInput

mcarrano commented:
@MariaAga Took a look at the preview and this is looking much better. Works great with the mouse, but didn't have much success entering a date or time directly from the keyboard. I get the text cursor bit cannot enter directly into those fields.

patternfly-build commented:
PatternFly-React preview: https://1420-pr-patternfly-react-patternfly.surge.sh

MariaAga commented:
Thanks @jeff-phillips-18, fixed all the issues you described.
@mcarrano Thanks for looking into it! I fixed this issue and now text is easier to input.

jeff-phillips-18 commented on packages/patternfly-3/patternfly-react/less/_date-time-picker.less:
```diff
@@ -0,0 +1,59 @@
+      span {
+        height: 24px;
+        line-height: 24px;
+        color: #0088ce;
```
color-pf-blue-400

jeff-phillips-18 commented on packages/patternfly-3/patternfly-react/less/_date-time-picker.less:
```diff
@@ -0,0 +1,59 @@
+  }
+  .timepicker-hours .table-condensed > tbody > tr > td,
+  .timepicker-minutes .table-condensed > tbody > tr > td {
+    padding: 7px 0px 7px 0px;
```
```suggestion
    padding: 7px 0;
```

jeff-phillips-18 commented on packages/patternfly-3/patternfly-react/less/_date-time-picker.less:
```diff
@@ -0,0 +1,59 @@
+  }
+  table td {
+    .today-button {
+      padding: 0px 16px 0px 0px;
```
```suggestion
      padding: 0 16px 0 0;
```

jeff-phillips-18 commented on packages/patternfly-3/patternfly-react/src/components/DateTimePicker/DateComponents/Header.js:
```diff
@@ -0,0 +1,67 @@
+        </tr>
+        <tr>
+          {daysOfTheWeek.map((day, idx) => (
+            <th key={idx} className="dow">
```
Is this a bootstrap class? It doesn't appear to have any values. I can see it being useful to applications to give access to the styling, but `dow` isn't all that intuitive to me.

MariaAga replied:
it was in the [patternfly code](https://www.patternfly.org/pattern-library/forms-and-controls/date-and-time/#code), should I remove it?

jeff-phillips-18 replied:
It's ok, it at least gives a hook. If anyone is moving from pf -> pf-react, they might expect this class to still be on the items.

jeff-phillips-18 commented:
@mcarrano when you are OK with the UX could you update the label please?

mcarrano commented:
Thanks for making the changes @MariaAga . I updated to 'UX Approved'

MariaAga commented:
Split this PR into 2 - one for the date picker, and one for the time picker 
https://github.com/patternfly/patternfly-react/pull/1875
https://github.com/patternfly/patternfly-react/pull/1873

jeff-phillips-18 commented:
Why the split? Should this PR be closed?

MariaAga commented:
Split for easier review of the code.
I would like to leave this PR open as this one includes the combination of the time picker and the date picker in the date time picker component

jeff-phillips-18 commented:
Should this then depend on the other two merging first?

MariaAga commented:
Yes

patternfly-build commented:
PatternFly-React preview: https://1420-pr-patternfly-react-patternfly.surge.sh

waldenraines commented:
@jeff-phillips-18 @tlabaj can someone review this please?

jeff-phillips-18 commented:
I believe this depends on #1873 , then some rework.

MariaAga commented:
Rebased

patternfly-build commented:
PatternFly-React preview: https://patternfly-react-pr-1420.surge.sh

MariaAga commented:
Updated the snapshot, thanks!

patternfly-build commented:
Your changes have been released in: 
 - patternfly-react-extensions@2.20.5
 - patternfly-react@2.39.0
 - @patternfly/react-console@1.12.9

Thanks for your contribution! :tada:
